server/test/test_repository.go

```diff
@@ -0,0 +1,236 @@
+package test
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/kubeshop/tracetest/server/pkg/id"
+)
+
+type Repository interface {
+	List(_ context.Context, take, skip int, query, sortBy, sortDirection string) ([]Test, error)
+	Count(context.Context, string) (int, error)
+	SortingFields() []string
+	Provision(context.Context, Test) error
+	SetID(test Test, id id.ID) Test
+
+	// TODO: uncomment as we add new functionality
+	// Get(context.Context, id.ID) (Test, error)
+	// Create(context.Context, Test) (Test, error)
+	// Update(context.Context, Test) (Test, error)
+	// Delete(context.Context, Test) error
+	// Exists(context.Context, id.ID) (bool, error)
+	// GetTestVersion(_ context.Context, _ id.ID, version int) (Test, error)
+	// List(_ context.Context, take, skip int32, query, sortBy, sortDirection string) ([]Test, error)
+}
+
+type repository struct {
+	db *sql.DB
+}
+
+func NewRepository(db *sql.DB) Repository {
+	return &repository{db}
+}
+
+func (r *repository) SetID(test Test, id id.ID) Test {
+	test.ID = id
+	return test
+}
+
+func (r *repository) Provision(ctx context.Context, test Test) error {
+	return nil
+}
+
+func (r *repository) SortingFields() []string {
+	return []string{"created", "name", "last_run"}
+}
+
+const (
+	getTestSQL = `
+	SELECT
+		t.id,
+		t.version,
+		t.name,
+		t.description,
+		t.service_under_test,
+		t.specs,
+		t.outputs,
+		t.created_at,
+		(SELECT COUNT(*) FROM test_runs tr WHERE tr.test_id = t.id) as total_runs,
+		last_test_run.created_at as last_test_run_time,
+		last_test_run.pass as last_test_run_pass,
+		last_test_run.fail as last_test_run_fail
+	FROM tests t
+	LEFT OUTER JOIN (
+		SELECT MAX(id) as id, test_id FROM test_runs GROUP BY test_id
+	) as ltr ON ltr.test_id = t.id
+	LEFT OUTER JOIN
+		test_runs last_test_run
+	ON last_test_run.test_id = ltr.test_id AND last_test_run.id = ltr.id
+`
+
+	testMaxVersionQuery = `
+	INNER JOIN (
+		SELECT id as idx, max(version) as latest_version FROM tests GROUP BY idx
+		) as latest_tests ON latest_tests.idx = t.id AND t.version = latest_tests.latest_version
+	`
+)
+
+func (r *repository) List(ctx context.Context, take, skip int, query, sortBy, sortDirection string) ([]Test, error) {
+	hasSearchQuery := query != ""
```

Comment:
we have some reusable sql functions in the `pkg/sqlutil` package. for example:

- search: https://github.com/kubeshop/tracetest/blob/main/server/tests/transactions.go#L433
- sort: https://github.com/kubeshop/tracetest/blob/main/server/tests/transactions.go#L328